mlir/lib/Dialect/Rock/Transforms/FoldBroadcast.cpp

```diff
@@ -0,0 +1,174 @@
+//===- Regularize.cpp - rewrites to allow Rock kernel fusion  ------===//
+//
+// Copyright 2022 Advanced Micro Devices.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Utils/Utils.h"
+
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/IR/RockGemmWrapperInterface.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/Support/Debug.h"
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKFOLDBROADCASTPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-canonicalize-gemm"
+
+using namespace mlir;
+using namespace mlir::rock;
+
+namespace {
+
+struct FoldBroadcast : public OpRewritePattern<rock::GemmOp> {
+  using OpRewritePattern<rock::GemmOp>::OpRewritePattern;
+
+  // Determine if the first dimension of a view is a broadcast
+  bool isBatchDimFoldable(PatternRewriter &rw, Value aView) const {
+    auto [buffer, views, needs64BitIdxA] = untransform(rw, aView);
+
+    // There are no views, hence no broadcast is possible
+    if (views.empty())
+      return false;
+
+    auto trMap = views[0].cast<TransformMapAttr>();
+
+    // There is no batch, hence nothing that can be a broadcast
+    if (trMap.getUpperBounds().size() != 3)
+      return false;
+
+    auto ops = trMap.getOps();
+
+    // Look for a `Broadcast{1} at [0]` transformation
+    for (auto op : ops)
+      if (op.getType() == rock::TransformType::Broadcast &&
+          op.getUpperDims().back() == 0 && op.getParams().back() == 1)
```

Comment:
Im a bit vary about this check... it is a bit fragile because unit dims / reshapes / transposes ? can easily break the ordering but it could be a still be a isBatchDimFoldable.

I think we need to backtrack a stride calculation on the batch dimension to see if its end up being zero on the actual memory layout.

Reply:
Just to be sure, you are saying that a broadcast might also not be upfront, right? It might be a passthrough/slice that leads to a broadcast.

Reply:
yes -- any other transform to be precise.

Reply:
I am not sure what you mean with "any other transform"? I was thinking to "rerouting" transforms, because I am not sure how can I deal with `broadcast` buried behind a Merge/unmerge/embed

Reply:
well the idea I had was to calculate stride of "g" dimension traversing up the trMaps.
Cant we handle the stride calculation in a merge/unmerge as the cumalative mulplication of faster changing dimensions in the merge set ? 

Embed is the tricky one but since we are only interested of figuring out stride is zero or not. We can possibly inherit the stride from bottom coords from the embed set

Reply:
I think it is quite plausible that we may encounter shape collapsing and expansion around the broadcast -- which usually result in merge/unmerge 

Embeds occur in conv to gemm I think. Thats a good point! we should test a convolution too.

Reply:
That can also work in most cases until batches are collapsed/expanded after the broadcast -- which I agree is unlikely.
I was more speaking about strides thing was : dont we need that for problem key v2 ?. 
Therefore if we have that, it is just a matter of looking at the stride of "g".

Reply:
Yep this was a plan I had, but I didn't do it just I didn't know if I was being too general. I think my confusion is about: 

> If the G dimension is created by Merge[*], Unmerge, or Embed, abort.

@manupak was that what you meant? Or you were suggesting to go through these transformations as well?

Reply:
So there is a reason we have the following code (it was done by Jungwook): 

https://github.com/ROCm/rocMLIR/blame/3202ef9b92a9278708b445a2611fff41750ffeea/mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp#L297-L321

to me it indicates that we will probably add a reshape to batch dimension and it would be after the broadcast because the above case handle where there are multiple dimensions (can just be unit dims) and fold them to be a single batch dimension.

Reply:
> @manupak is that what you meant? Or you were suggesting to go through these transformations as well?

I was suggesting to calculate/update a stride as we go through the stack and stop if and when it becomes zero (i.e. broadcast).

I ll point out that @krzysz00 's logic will work until we have a reshape that involves batch dimension after the broadcast. Now, thinking about it in the broadcast case, we might not commonly exercise that case that but I am a bit worried about unit dims.

Anyway, when we have problem key v2, we will have to calculate the strides. Therefore, whatever logic we are discussing could just go away and we can just take the decision looking at the stride ?